### PR TITLE
Narrative fixes and improvements

### PR DESCRIPTION
**Pull request by eapearson**
# Search Indexer Changes

## Kafka Producer Misses

This set of changes was triggered by frequent producer misses (messages were sent, but never received). Updating from 0.11.6 to 1.3.0 improved the miss rate, but did not solve it.

What did solve it was rearranging the kafka client calls. Specifically, I suspect that the `_produce` method in the main index runner was doing two things wrong. First, it was creating a Kafka producer client in the method, and then calling `poll` with a 1/10 second timeout. This has the effect of potentially not finalizing the message send (if it takes more than 1/10 second), and perhaps not even sending the message, considering that the lifetime of the producer client object was the method. The producer client maintains an internal queue of messages to send -- destroying the client (or rather having it fall out of scope and be garbage collected some time in the near future) before it can send them may result in the loss of messages.

Finally, the usage of a 1/10 second timeout in the poll call means that indexing time can be impacted as each call may wait for up to 1/10 sec for the message to be delivered and reported. (This may not really be an issue, as indexing time is dominated by io fetching objects.)

So the changes were:

- update the kafka client
- create a producer client which has a lifetime of the index runner, not the method
- calling producer.poll with 0 timeout in the _produce method
- handling buffer full errors by polling with a timeout in a short loop
- call producer.poll with 0 timeout in the main kafka receive loop; that loop is infinite, so it is sure to send the messages eventually. Another option would be a separate loop (async, thread?) just for producer polling, but seems unnecessary - that is a common technique for long-running apps to manage outgoing queues.)
- refactored kafka shutdown into a single method which closes both the consumer and producer; the producer shutdown drains the queue (sends all waiting messages.)

> TODO: currently that code enters an infinite loop, until the message count is 0 or an error occurs.

## Reformatting

I didn't reformat code intentionally, but I do have pyflakes8 running, which will reformat code on save, and the redhat yaml linter as well.

## Skipping Relation Engine

I did not want to deal with any relation engine business, but when I attempted to use `SKIP_RELENG` I found that the relation engine code was not fully avoided. There are several changes related to ensuring that using SKIP_RELENG works as expected.

## Support non-admin token

There are a bunch of changes to support using a token without workspace admin privileges.

This is important to allow developers to more easily contribute to the search indexer, api, and specs. Admin access to workspaces should be limited. Removing the requirement to obtain workspace admin privs, eases additional developer contribution.

Changes include:

- a new environment variable, `WS_ADMIN`, which defaults to "yes"
- moving all workspace client code into ws_utils, which provides methods which mirror the non-admin method names, and internally dispatch to either the admin or non-admin methods, depending on the WS_ADMIN setting
- created CI user `kbasesearchtest1` for usage in local dev indexing and testing (in searchapi2); we should discuss. We use a similar approach in kbase-ui integration testing (`kbaseuitest`).

## Local Developer Tools

A new directory, local-dev, contains docker-compose configs, instructions, and scripts for running the indexer locally, and populating it with data from a kbase deploy environment.

## Optimization: don't fetch data when object skipped

The code to handle an index request was fetching the entire object for each requested object index operation, even if the object was not going to be indexed, due to blacklisting or type filtering. This signficantly speeded up indexing when indexes are whitelisted (`ALLOW_INDICES`)

## Optimization: don't get object v1 if that is the only one

The code to get the first version of an object in order to determine the creation time was optmized for objects with only a single version. In that case, the v1 object is not fetched, rather the object already fetched is used.

## Option to omit large objects

When working locally, indexing large objects (> 100M or so) can severely impact one's workstation, and even crash the indexer (I'm assuming OOM).

A simple measure for local development is to simply skip indexing (and thus fetching, parsing, etc.) very large objects.

A new environment variable option `MAX_OBJECT_SIZE` is provided for this purpose. It is `None` by default, which skips the test for large object skipping.

## Narrative index improvements/fixes

The narrative indexer was altered to add additional fields:

- owner
- is_temporary
- is_narratorial
- modified_at (workspace modified, not narrative object)

Also added error messages ('_action': 'error') for various error conditions - missing metadata, messing `is_temporary` metadata field.

## Load config from file or url

In order to work on a search index, the search spec also needs altering. The current spec loader works for urls only. Now a spec config is looked for in `/app/config/config.yaml` first. This file can be placed there by a docker volume mount, as is done in the local developer scripts.

If the file is not present, the spec is fetched from github as usual.

## Moved ES related functions to es_utils

A few private functions which interact with ES directly, like `_log_err_to_es`,  were moved into `es_utils`.

## Test Changes

Tests initially failed due to a race condition between starting the service and the test runner. In addition, the addition of non-admin mode means that mock test data needed additions.

### Test startup fixed

The test process (`make test`) starts all prerequisite services, and then the service is started in the background, and then the tests are run. However, resource constraints on my laptop revealed that when the service is slow to start, the tests can start before the server is ready and the tests fail.

The server slow start is partially due to checking for the readiness of ES and RE. 

So the solution is to ensure that the tests only start running after the service itself is ready. The implementation chosen is for the service to create and update a state file, which an external script can monitor, pausing execution of the test script until the state file is in the correct state (containing the text "ready"), and then continuing if the server is ready, or failing if a time limit is exceeded.

A new script `ensure-indexrunner-read.py` is invoked by `run_tests.sh` to poll the status of the ready file, and exit if the index runner is not ready within a time limit. This time limit is controlled by the `READY_TIMEOUT` env var, which defaults to 60 (seconds).

### Test data updated

In order to test for usage with a non-admin token, the workspace mock data needed to be duplicated and altered to match the non-admin calls (which are normal workspace api calls) - just the param signature changed.

> Note: the object versions used in the test data have been deleted from CI, although the object with a later version has not been.

### local spec usage

As a nod to local developer usage, the indexer checks to see if there is a local spec file before fetching. This allows a developer to volume mount the specs repo into the indexer container in order to iterate locally on the specs.

## TODO and/or Discuss

- usage of and management of `kbasesearchtestX` users
- generation of test data from CI (e.g. current object version used in test data  has been deleted)
- type blacklist should probably be in the spec repo, but is in es_indexers/main.py
- should releng indexing be separated?
- it isn't clear to me how the entanglement of indexing code in the indexer, the specs, and indexing modules are kept in sync other than carefully synced releases.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,5 +1,6 @@
 # Environment variables
-.env
+# why .env was blacklisted? it is useful for the docker-compose.yaml testing file.
+# .env
```
Comment (scanon) [inline .gitignore]: .env should be blacklisted and the .env above should be called env.example or something similar.

File: local-dev/indexing-scripts/index-mycocosm.rb [hunk 2]
```diff
@@ -0,0 +1,42 @@
+require './services'
```
Comment (scanon) [inline local-dev/indexing-scripts/index-mycocosm.rb]: Why are we using ruby for these test scripts.  Why introduce yet another language in the middle of this?

File: local-dev/README.md [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+Well, that is a lie. There are object-level filters for:
+
+- object size (`MAX_OBJECT_SIZE`)
+- blacklisted types (defined the indexer spec)
+
+The scripts are easy enough to understand that if you have a special case, it should be feasible to create a custom indexing script. If it is generally useful, check it in!
+
+### Using the scripts
+
+The scripts are, at present, implemented in Ruby. All you need is a relatively recent version of Ruby 2.X. No external dependencies are required.
```
Comment (scanon) [inline local-dev/README.md]: See my notes below. I would prefer not to add ruby to the mix here.  Can't these be done in bash or python?

File: local-dev/scripts/run_narratives.sh [hunk 4]
```diff
@@ -0,0 +1,22 @@
+cd ..
```
Comment (scanon) [inline local-dev/scripts/run_narratives.sh]: I don't understand the purpose of this and why it is called run_narratives.

File: local-dev/scripts/build.sh [hunk 5]
```diff
@@ -0,0 +1,3 @@
+cd ..
```
Comment (scanon) [inline local-dev/scripts/build.sh]: This could probably just be in the Makefile

File: src/utils/ws_utils.py [hunk 6]
```diff
@@ -1,3 +1,9 @@
+from src.utils.config import config
```
Comment (scanon) [inline src/utils/ws_utils.py]: I think this should be a class and it should basically mimic the standard workspace client calls.

File: src/utils/es_utils.py [hunk 7]
```diff
@@ -1,8 +1,10 @@
 """Elasticsearch API client utilities."""
 import json
 import requests
+import hashlib
 
 from src.utils.config import config
+import src.index_runner.es_indexer as es_indexer
```
Comment (scanon) [inline src/utils/es_utils.py]: This introduces a circular dependency between es_utils and es_indexer.  That seems like a bad idea.  We should figure out what belongs where.

File: src/index_runner/es_indexers/main.py [hunk 8]
```diff
@@ -32,52 +30,74 @@ def index_obj(obj_data, ws_info, msg_data):
...
+    ws_metadata = ws_info[-1]
+    # If the workspace's object metadata contains a "noindex" tag, skip it
+    if ws_metadata.get('searchtags'):
+        if 'noindex' in ws_metadata['searchtags']:
+            logger.debug(f'Skipping "noindex" workspace object {ws_info[0]}, {obj_info[6]}')
             return
+
+    # We may refuse to index very large objects
+    # TODO: should be very high in prod, lower for development
+    max_object_size = config().get('max_object_size')
```
Comment (scanon) [inline src/index_runner/es_indexers/main.py]: I think we should construct the indexes to accommodate for this.  I can't think of an instance where we would want this to happen.

File: .gitignore (hunk 1 above)
Comment (jayrbolton) [reply .gitignore]: The purpose of docker-compose.yaml is testing, so it can contain all testing env vars directly

**Pull request closed without merging**